bellman/src/domain/fft.rs

```diff
@@ -0,0 +1,177 @@
+    for k in 0..len {
+        let rk: u32 = bitreverse(k, log_n);
+        if k < rk {
+            series.swap(rk as usize, k as usize);
+        }
+    }
+
+    let mut m: u32 = 1;
+    for _ in 0..log_n {
+        let w_m = omega.pow(&[(len / (2 * m)) as u64]);
+
+        let mut k = 0;
+        while k < len {
+            let mut w: _ = E::Fr::one();
+            for j in 0..m {
+                let mut t: T = series[(k + j + m) as usize];
+                t *= &w;
+
+                let mut tmp: T = series[(k + j) as usize];
+                tmp -= &t;
+
+                series[(k + j + m) as usize] = tmp;
+                series[(k + j) as usize] += &t;
+
+                w.mul_assign(&w_m);
+            }
+
+            k += 2 * m;
+        }
+
+        m *= 2;
+    }
+}
+
+fn bitreverse(mut n: u32, l: u32) -> u32 {
+    let mut r = 0;
+    for _ in 0..l {
+        r = (r << 1) | (n & 1);
+        n >>= 1;
+    }
+    r
+}
+
+pub fn parallel_fft<E,T>(a: &mut [T], worker: &Worker, omega: &E::Fr, log_n: u32, log_cpus: u32) 
+where
+    E: ScalarEngine,
+    for <'a> T: Group<'a,E>
+{
+    assert!(log_n >= log_cpus);
+
+    let num_cpus = 1 << log_cpus;
+    let log_new_n = log_n - log_cpus;
+    let mut tmp = vec![vec![T::zero(); 1 << log_new_n]; num_cpus];
+    let new_omega = omega.pow(&[num_cpus as u64]);
+
+    worker.scope(0, |scope, _| {
+        let a: &[T] = a;
+
+        for (j, tmp) in tmp.iter_mut()
+            .enumerate() 
+        {
+            scope.spawn(move || {
+                // Shuffle into a sub-FFT
+                let omega_j: _ = omega.pow(&[j as u64]);
+                let omega_step: _ = omega.pow(&[(j as u64) << log_new_n]);
+
+                let mut elt: _ = E::Fr::one();
+                for i in 0..(1 << log_new_n) {
+                    for s in 0..num_cpus {
+                        let idx = (i + (s << log_new_n)) % (1 << log_n);
+                        let mut t = a[idx];
+                        t *= &elt;
+                        tmp[i] += &t;
+                        elt.mul_assign(&omega_step);
+                    }
+                    elt.mul_assign(&omega_j);
+                }
+
+                // Perform sub-FFT
+                serial_fft(tmp, &new_omega, log_new_n);
+            });
+        }
+    });
+
+    // TODO: does this hurt or help?
+    worker.scope(a.len(), |scope, chunk| {
+        let tmp = &tmp;
+
+        for (idx, a) in a.chunks_mut(chunk).enumerate() {
+            scope.spawn(move || {
+                let mut idx = idx * chunk;
+                let mask = (1 << log_cpus) - 1;
+                for a in a {
+                    *a = tmp[idx & mask][idx >> log_cpus];
+                    idx += 1;
+                }
+            });
+        }
+    });
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use arith::Scalar;
+    use crate::domain::Domain;
+
+    #[cfg(feature = "pairing")]
+    #[test]
+    fn fft_composition() {
+        use pairing::bls12_381::Bls12;
```

Comment:
Please look into Arrange/Act/Assert pattern to order tests. All tests/helper functions should state clearly what they do and this does not.

Reply:
Tests have been left out for this refactor to avoid accidentally breaking them for the moment, but will keep in mind when I do take them on